Add FFT support via AbstractFFTs interface

KaanKesginLW opened the pull request:
Adds FFT support for `MtlArray` via the AbstractFFTs.jl interface.

HEAVILY based on CUDA.jl's AbstractFFTs.jl interface implementation using MPSGraph functionality.

```julia
using Metal

x = MtlArray(randn(ComplexF32, 2048, 2048))
y = fft(x)  # Just works!
```

### Performance

Benchmarked on Apple M2 Max with 30-core GPU against FFTW.jl on CPU:

| Size | CPU (FFTW)  | CPU (FFTW & AppleAccelerate) | GPU (Metal) |
|------|------------|-------------
| 512×512 | 4.2ms | 766.2μs | 173.4μs |
| 1024×1024 | 19.7ms | 3.7ms | 246.3μs |
| 2048×2048 | 99.5ms | 20.8ms | 588.4μs |
| 4096×4096 | 580.1ms | 99.0ms | 2.5ms |

### Example Usage

```julia
using Metal

# Complex FFT
x = MtlArray(randn(ComplexF32, 1024, 1024))
y = fft(x)
z = ifft(y)  # z ≈ x

# Real FFT  
r = MtlArray(randn(Float32, 1024, 1024))
c = rfft(r)           # Real → Complex
r2 = irfft(c, 1024)   # Complex → Real, r2 ≈ r

# FFT along specific dimensions
y = fft(x, 1)         # First dimension only
y = fft(x, (1, 2))    # Batched transform

# Plan reuse
x = MtlArray(randn(ComplexF32, 1024, 1024))
another_x = MtlArray(randn(ComplexF32, 1024, 1024))
p = plan_fft(x)
y1 = p * x
y2 = p * another_x    # Same plan, different data
```

Close #270

KaanKesginLW committed:
Add FFT support via AbstractFFTs interface

Implements GPU FFT operations for MtlArray using MPSGraph's
fastFourierTransformWithTensor. This addresses issue #270.

Features:
- plan_fft, plan_ifft, plan_bfft for ComplexF32 arrays
- Multi-dimensional FFT support (single axis, multiple axes, all axes)
- FFTW.jl-compatible API via AbstractFFTs.jl interface
- Plan execution via * operator and mul!

Implementation notes:
- Uses MPSGraphFFTDescriptor for forward/inverse control
- Scaling handled manually for multi-axis ifft (not via MPSGraph's
  scalingMode) to ensure correct normalization across all FFT dimensions
- Axis mapping accounts for Julia's column-major vs Metal's row-major
  ordering via shape reversal in placeholderTensor

Tested against FFTW.jl with <1e-4 relative tolerance for all operations.

KaanKesginLW committed:
Add FFT support via AbstractFFTs interface

Implements GPU FFT operations for MtlArray using MPSGraph's
fastFourierTransformWithTensor. This addresses issue #270.

Features:
- plan_fft, plan_ifft, plan_bfft for ComplexF32 arrays
- Multi-dimensional FFT support (single axis, multiple axes, all axes)
- FFTW.jl-compatible API via AbstractFFTs.jl interface
- Plan execution via * operator and mul!

Implementation notes:
- Uses MPSGraphFFTDescriptor for forward/inverse control
- Scaling handled manually for multi-axis ifft (not via MPSGraph's
  scalingMode) to ensure correct normalization across all FFT dimensions
- Axis mapping accounts for Julia's column-major vs Metal's row-major
  ordering via shape reversal in placeholderTensor

Tested against FFTW.jl with <1e-4 relative tolerance for all operations.

KaanKesginLW committed:
Add real FFT support (rfft, irfft, brfft)

Implements real-to-complex and complex-to-real FFT operations using
MPSGraph's realToHermiteanFFTWithTensor and HermiteanToRealFFTWithTensor.

Features:
- plan_rfft for Float32 arrays (output size n÷2+1 in first FFT dimension)
- plan_irfft for ComplexF32 arrays (normalized inverse)
- plan_brfft for ComplexF32 arrays (unnormalized inverse)
- Proper handling of odd output sizes via roundToOddHermitean
- FFTW-compatible dimension conventions

The output size reduction follows FFTW convention: the first transformed
dimension is reduced to n÷2+1 for rfft, and irfft requires the original
size to be specified.

Tested against FFTW.jl with <1e-4 relative tolerance.

KaanKesginLW committed:
Add real FFT support (rfft, irfft, brfft)

Implements real-to-complex and complex-to-real FFT operations using
MPSGraph's realToHermiteanFFTWithTensor and HermiteanToRealFFTWithTensor.

Features:
- plan_rfft for Float32 arrays (output size n÷2+1 in first FFT dimension)
- plan_irfft for ComplexF32 arrays (normalized inverse)
- plan_brfft for ComplexF32 arrays (unnormalized inverse)
- Proper handling of odd output sizes via roundToOddHermitean
- FFTW-compatible dimension conventions

The output size reduction follows FFTW convention: the first transformed
dimension is reduced to n÷2+1 for rfft, and irfft requires the original
size to be specified.

Tested against FFTW.jl with <1e-4 relative tolerance.

KaanKesginLW committed:
Add ComplexF16/Float16 support and improve type documentation

Phase 4: Type generalization for FFT operations.

Changes:
- Add ComplexF16 support for fft/ifft/bfft
- Add Float16 support for rfft/irfft/brfft
- Improve error messages for unsupported types (ComplexF64/Float64)
- Add documentation about supported and unsupported types
- Add FFTComplexTypes and FFTRealTypes type unions

Note: ComplexF64/Float64 are NOT supported by Metal's MPSGraph FFT.
Users requiring double precision should use FFTW.jl on CPU.

ComplexF16 results have ~3 decimal digits precision (expected for Float16).

KaanKesginLW committed:
Add ComplexF16/Float16 support and improve type documentation

Phase 4: Type generalization for FFT operations.

Changes:
- Add ComplexF16 support for fft/ifft/bfft
- Add Float16 support for rfft/irfft/brfft
- Improve error messages for unsupported types (ComplexF64/Float64)
- Add documentation about supported and unsupported types
- Add FFTComplexTypes and FFTRealTypes type unions

Note: ComplexF64/Float64 are NOT supported by Metal's MPSGraph FFT.
Users requiring double precision should use FFTW.jl on CPU.

ComplexF16 results have ~3 decimal digits precision (expected for Float16).

KaanKesginLW committed:
Add 1D FFT tests

Phase 5: Verify and test 1D FFT support.

The existing implementation already handles 1D arrays correctly since
the multi-dimensional FFT code works for arbitrary dimensions.

Added tests:
- 1D fft correctness vs FFTW
- 1D ifft roundtrip
- 1D rfft correctness vs FFTW
- 1D rfft -> irfft roundtrip

KaanKesginLW committed:
Add 1D FFT tests

Phase 5: Verify and test 1D FFT support.

The existing implementation already handles 1D arrays correctly since
the multi-dimensional FFT code works for arbitrary dimensions.

Added tests:
- 1D fft correctness vs FFTW
- 1D ifft roundtrip
- 1D rfft correctness vs FFTW
- 1D rfft -> irfft roundtrip

KaanKesginLW committed:
Add in-place FFT variants (fft!, ifft!, bfft!)

Phase 6: Implement in-place FFT operations.

New types:
- MtlFFTInplacePlan{T,K,N} - plan that modifies input directly

New functions:
- plan_fft!(x, [region]) - create in-place forward FFT plan
- plan_ifft!(x, [region]) - create in-place inverse FFT plan
- plan_bfft!(x, [region]) - create in-place backward FFT plan

The in-place variants are useful for avoiding memory allocation when
the input data is no longer needed after the transform.

Usage:
    x = MtlArray(randn(ComplexF32, 64, 64))
    plan = plan_fft!(x)
    plan * x  # x is modified in-place, returns x

KaanKesginLW committed:
Add in-place FFT variants (fft!, ifft!, bfft!)

Phase 6: Implement in-place FFT operations.

New types:
- MtlFFTInplacePlan{T,K,N} - plan that modifies input directly

New functions:
- plan_fft!(x, [region]) - create in-place forward FFT plan
- plan_ifft!(x, [region]) - create in-place inverse FFT plan
- plan_bfft!(x, [region]) - create in-place backward FFT plan

The in-place variants are useful for avoiding memory allocation when
the input data is no longer needed after the transform.

Usage:
    x = MtlArray(randn(ComplexF32, 64, 64))
    plan = plan_fft!(x)
    plan * x  # x is modified in-place, returns x

KaanKesginLW committed:
Apply Runic formatting to FFT implementation

KaanKesginLW committed:
Apply Runic formatting to FFT implementation

KaanKesginLW committed:
Add missing mul! import in FFT tests

KaanKesginLW committed:
Add missing mul! import in FFT tests

KaanKesginLW committed:
Add fused fftshift support for FFT operations

Add shift keyword argument to FFT plans and convenience functions
that fuses fftshift/ifftshift into the GPU graph execution.

- plan_fft(x; shift=true) fuses fftshift after forward FFT
- plan_ifft(x; shift=true) fuses ifftshift before inverse FFT
- Convenience functions in Metal.MPSGraphs namespace support shift kwarg
- Implemented via slice+concat in MPSGraph (no intermediate memory ops)
- Handles both even and odd array sizes correctly

Example usage:
  # Via plans
  p = plan_fft(x_gpu; shift=true)
  result = p * x_gpu

  # Via convenience functions
  using Metal.MPSGraphs: fft, ifft
  result = fft(x_gpu; shift=true)

KaanKesginLW committed:
Add fused fftshift support for FFT operations

Add shift keyword argument to FFT plans and convenience functions
that fuses fftshift/ifftshift into the GPU graph execution.

- plan_fft(x; shift=true) fuses fftshift after forward FFT
- plan_ifft(x; shift=true) fuses ifftshift before inverse FFT
- Convenience functions in Metal.MPSGraphs namespace support shift kwarg
- Implemented via slice+concat in MPSGraph (no intermediate memory ops)
- Handles both even and odd array sizes correctly

Example usage:
  # Via plans
  p = plan_fft(x_gpu; shift=true)
  result = p * x_gpu

  # Via convenience functions
  using Metal.MPSGraphs: fft, ifft
  result = fft(x_gpu; shift=true)

KaanKesginLW committed:
Simplify fused fftshift API: extend AbstractFFTs.fft directly

Allow fft(x; shift=true) syntax by extending AbstractFFTs.fft with
a shift keyword argument for MtlArray inputs. This provides a cleaner
API without requiring separate imports.

KaanKesginLW committed:
Simplify fused fftshift API: extend AbstractFFTs.fft directly

Allow fft(x; shift=true) syntax by extending AbstractFFTs.fft with
a shift keyword argument for MtlArray inputs. This provides a cleaner
API without requiring separate imports.

christiangnrd commented:
Thank you for your PR, I know many people have been waiting for someone to get around to implementing this for a while!

This is a great start! Before I fully dive in to review, here are a few changes I'd like you to make to simplify review (and future maintainance) of the code:
- All `@objc` calls should be wrapped so that everything else can be done with Julia code. Graph operations can go in lib/mpsgraph/operations.jl, and constructors can go at the top of the fft.jl file.
- If you could structure the implementation more closely with [CUDA's implementation](https://github.com/JuliaGPU/CUDA.jl/tree/master/lib/cufft), it would make it quicker for me to revie
- Similarly to the last point, for tests, please adapt the CUDA fft tests to here, so that we can take advantage of their regressions tests. Feel free to add tests from the current suite that the CUDA tests don't cover.

Feel free to ask if you have any questions.

KaanKesginLW committed:
Refactor FFT: move @objc wrappers to operations.jl

- Add sliceTensor, concatTensors to operations.jl
- Add fastFourierTransformWithTensor to operations.jl
- Add realToHermiteanFFTWithTensor, HermiteanToRealFFTWithTensor to operations.jl
- Remove duplicate wrapper functions from fft.jl
- Update _apply_fftshift_to_tensor to use operations.jl functions

KaanKesginLW committed:
Refactor FFT: move @objc wrappers to operations.jl

- Add sliceTensor, concatTensors to operations.jl
- Add fastFourierTransformWithTensor to operations.jl
- Add realToHermiteanFFTWithTensor, HermiteanToRealFFTWithTensor to operations.jl
- Remove duplicate wrapper functions from fft.jl
- Update _apply_fftshift_to_tensor to use operations.jl functions

KaanKesginLW committed:
Simplify FFT: remove fused fftshift feature

- Remove shift parameter from all plan types and functions
- Remove _apply_fftshift_to_tensor helper
- Simplify _execute_single_axis_fft!
- Remove fftshift tests (14 tests removed)
- Aligns API with CUDA.jl which has no built-in fftshift

KaanKesginLW committed:
Simplify FFT: remove fused fftshift feature

- Remove shift parameter from all plan types and functions
- Remove _apply_fftshift_to_tensor helper
- Simplify _execute_single_axis_fft!
- Remove fftshift tests (14 tests removed)
- Aligns API with CUDA.jl which has no built-in fftshift

KaanKesginLW committed:
Restructure FFT tests following CUDA.jl pattern

- Add Float16 CPU shims for FFTW reference comparisons
- Add tolerance functions based on type precision
- Restructure tests with reusable test functions
- Add @inferred checks for plan creation
- Add comprehensive batched transform tests (1D, 2D, 3D)
- Test both ComplexF16/ComplexF32 and Float16/Float32
- Tests increased from 55 to 117

KaanKesginLW committed:
Restructure FFT tests following CUDA.jl pattern

- Add Float16 CPU shims for FFTW reference comparisons
- Add tolerance functions based on type precision
- Restructure tests with reusable test functions
- Add @inferred checks for plan creation
- Add comprehensive batched transform tests (1D, 2D, 3D)
- Test both ComplexF16/ComplexF32 and Float16/Float32
- Tests increased from 55 to 117

KaanKesginLW commented:
Thanks for the feedback! I've addressed all the requested changes:

1. **Wrapped @objc calls** — Graph operations (`fastFourierTransformWithTensor`, `realToHermiteanFFTWithTensor`, `HermiteanToRealFFTWithTensor`, `sliceTensor`, `concatTensors`) moved to `operations.jl`. Constructor kept at top of `fft.jl`.

2. **Aligned with CUDA.jl structure** — Removed the fused fftshift feature to match CUDA.jl's API. Simplified the implementation.

3. **Adapted CUDA.jl tests** — Added Float16 shims, tolerance functions, reusable test functions, `@inferred` checks, and batched transform tests. Test count: 117.

Ready for review!

KaanKesginLW committed:
Apply Runic formatting to FFT tests

KaanKesginLW committed:
Apply Runic formatting to FFT tests

KaanKesginLW committed:
Address review feedback: remove .gitignore change, restore original formatting in operations.jl

KaanKesginLW committed:
Address review feedback: remove .gitignore change, restore original formatting in operations.jl

christiangnrd committed:
Remove hallucinated kwargs

christiangnrd committed:
Remove hallucinated kwargs

christiangnrd committed:
Add more CUDA tests

christiangnrd committed:
Add more CUDA tests

christiangnrd reviewed (changes requested):
Hey @KaanKesginLW, sorry for taking so long to actually review, I had to find some time where I could dive into this while being somewhat unfamiliar with FFTs.

I rebased and fixed up the tests, so no need to touch those, but it seems like your code implements the interface 3 times. I would much rather only one implementation similar to how [CUDA.jl](https://github.com/JuliaGPU/CUDA.jl/blob/master/lib/cufft/fft.jl) does it.

christiangnrd committed:
Temp disable some new tests

christiangnrd committed:
Temp disable some new tests

christiangnrd committed:
[NFC] Reorganize code

christiangnrd committed:
[NFC] Reorganize code

christiangnrd committed:
[NFC] Cleanup

christiangnrd committed:
[NFC] Cleanup

christiangnrd committed:
Rename to `MPSGraphFFTDescriptor`

christiangnrd committed:
Rename to `MPSGraphFFTDescriptor`

christiangnrd committed:
assert_applicable

christiangnrd committed:
assert_applicable

christiangnrd committed:
Add inplace stuff

christiangnrd committed:
Add inplace stuff

christiangnrd committed:
Merge MtfFFTPlan and MtlFFTInplacePlan

christiangnrd committed:
Merge MtfFFTPlan and MtlFFTInplacePlan

christiangnrd committed:
Move some code around

christiangnrd committed:
Move some code around

christiangnrd committed:
Remove unecessary helper function

christiangnrd committed:
Remove unecessary helper function

christiangnrd committed:
Add output type parameter

christiangnrd committed:
Add output type parameter

christiangnrd committed:
Big merge

christiangnrd committed:
Big merge

christiangnrd committed:
Fix show

christiangnrd committed:
Fix show

christiangnrd committed:
Metaprogram ALL the things

christiangnrd committed:
Metaprogram ALL the things

christiangnrd committed:
Minor refactor

christiangnrd committed:
Minor refactor

christiangnrd committed:
Bring implementation more in line with the AbstractFFTs interface

christiangnrd committed:
Bring implementation more in line with the AbstractFFTs interface

christiangnrd committed:
Add some missing CUDA stuff

christiangnrd committed:
Add some missing CUDA stuff

christiangnrd committed:
autoreleasepool tweaks

christiangnrd committed:
autoreleasepool tweaks

christiangnrd committed:
More validation + fixup tests

christiangnrd committed:
More validation + fixup tests

christiangnrd committed:
Rename stuff for consistency with CUDA fft test

christiangnrd committed:
Rename stuff for consistency with CUDA fft test

christiangnrd committed:
Integer FFTs

christiangnrd committed:
Integer FFTs

christiangnrd committed:
Remove unused MPSGraph operations

They can be added in a different PR

christiangnrd committed:
Remove unused MPSGraph operations

They can be added in a different PR

christiangnrd committed:
Cleanup

christiangnrd committed:
Cleanup

christiangnrd committed:
Comment cleanup

christiangnrd committed:
Comment cleanup

christiangnrd committed:
Cache plans

Claude was instructed to implement this like the MPSGraphs matmul
caching

christiangnrd committed:
Cache plans

Claude was instructed to implement this like the MPSGraphs matmul
caching

christiangnrd commented on test/fft.jl:
```diff
@@ -0,0 +1,470 @@
+    y = similar(x, Float32)
+    WrappedFloat16Operator(plan_rfft(y, dims...))
+end
+function AbstractFFTs.plan_irfft(x::Array{Complex{Float16}}, dims...)
+    y = similar(x, Complex{Float32})
+    WrappedFloat16Operator(plan_irfft(y, dims...))
+end
+function AbstractFFTs.plan_brfft(x::Array{Complex{Float16}}, dims...)
+    y = similar(x, Complex{Float32})
+    WrappedFloat16Operator(plan_brfft(y, dims...))
```
Only affects tests and they're passing so skipping

christiangnrd commented on src/fft.jl:
```diff
@@ -0,0 +1,333 @@
+
+function plan_brfft(x::MtlArray{T, N}, d::Int, region::NTuple{R, Int}) where {T <: FFTComplex, N, R}
+    backward = true
+    inplace = false
```
Implementing

christiangnrd commented on src/fft.jl:
```diff
@@ -0,0 +1,333 @@
+
+# forward plans are `plan_fft`, and backward (unnormalized) plans are `plan_bfft`
+# inplace functions have a "!", inverse (normalized) plans are handled via plan_inv
+for inplace in (true, false), backward in (true, false)
+    dir_str = backward ? "b" : ""
+    inplace_str = inplace ? "!" : ""
+    f = Symbol(:plan_, dir_str, :fft, inplace_str)
+
+    @eval begin
+        # untyped `region` argument
+        Base.@constprop :aggressive function $f(x::MtlArray{T, N}, region) where {T <: FFTComplex, N}
+            R = length(region)
+            region = NTuple{R,Int}(region)
+            $f(x, region)
+        end
+
+        # actually create the MtlFFTPlan
+        $f(x::MtlArray{T, N}, region::NTuple{R, Int}) where {T <: FFTComplex, N, R} = MtlFFTPlan{T, T, $backward, $inplace, N, R}(size(x), size(x), region)
+    end
+end
```
This is just wrong

christiangnrd commented on src/fft.jl:
```diff
@@ -0,0 +1,333 @@
+
+GPU FFT plan for Metal using MPSGraph's fastFourierTransformWithTensor.
+"""
+mutable struct MtlFFTPlan{T <: FFTNumber, S <: FFTNumber, backward, inplace, N, R} <: Plan{S}
+    input_size::NTuple{N, Int}
+    output_size::NTuple{N, Int}
+    region::NTuple{R, Int}
+    pinv::ScaledPlan{T}
+
+    function MtlFFTPlan{T, S, backward, inplace, N, R}(input_size::NTuple{N, Int}, output_size::NTuple{N, Int}, region::NTuple{R, Int}) where {T <: FFTNumber, S <: FFTNumber, backward, inplace, N, R}
+        # Validate region
+        if any(diff(collect(region)) .< 1)
+            throw(ArgumentError("region must be an increasing sequence"))
+        end
+        if any(region .< 1 .|| region .> N)
+            throw(ArgumentError("region can only refer to valid dimensions"))
+        end
+        backward isa Bool || throw(ArgumentError("FFT backward argument must be a Bool"))
+        inplace isa Bool || throw(ArgumentError("FFT inplace argument must be a Bool"))
+
+        return new{T, S, backward, inplace, N, R}(input_size, output_size, region)
+    end
+end
```
Things work and this is how CUDA.jl does it.

christiangnrd committed:
Copilot suggestion

christiangnrd committed:
Copilot suggestion

christiangnrd commented:
@maleadt [This](https://buildkite.com/julialang/metal-dot-jl/builds/2631/steps/table?jid=019d7e5d-44af-4af6-8af5-2920b1067d4b&tab=output) failure happened because AppleAccelerate.jl was loaded in the flopscomp example test which heppened to be on the same runner as the fft tests. Should we just remove AppleAccelerate?

Also, Copilot seems to think that Dict reads while writes are potentially happening isn't thread safe. If that's true we'll need to fix the MPSGraph caching too.

maleadt commented:
> @maleadt [This](https://buildkite.com/julialang/metal-dot-jl/builds/2631/steps/table?jid=019d7e5d-44af-4af6-8af5-2920b1067d4b&tab=output) failure happened because AppleAccelerate.jl was loaded in the flopscomp example test which heppened to be on the same runner as the fft tests. Should we just remove AppleAccelerate?

Yeah I guess. It's unfortunate that AppleAccelerate breaks stuff.



> Also, Copilot seems to think that Dict reads while writes are potentially happening isn't thread safe. If that's true we'll need to fix the MPSGraph caching too.

Yes, Dict isn't threadsafe. You'd want to wrap those concurrent accesses in a ReentrantLock.

christiangnrd committed:
Fix potentially unsafe cache access

christiangnrd commented:
> Is there anything I can help with here?

This is mostly waiting for a review from someone familiar with the AbstractFFTs.jl interface since neither @maleadt or I have much experience with it.

giordano commented on test/fft.jl:
```diff
@@ -0,0 +1,470 @@
+atol(::Type{Float64}) = 1.0e-15
+atol(::Type{I}) where {I<:Integer} = atol(Metal.mtlfloat(I))
+rtol(::Type{Complex{T}}) where {T} = rtol(T)
+atol(::Type{Complex{T}}) where {T} = atol(T)
```
Absolute tolerance based on floating point type isn't very useful: the absolute tolerance depends on the scale of the numbers involved, not their "precision", and should in general be 0 unless the compared numbers are themselves 0 or extremely close to it.  With these definitions you get results like
```julia-repl
julia> isapprox(1e-15, 2e-15; rtol=rtol(Float64), atol=atol(Float64))
true

julia> isapprox(1e-15, 1e-16; rtol=rtol(Float64), atol=atol(Float64))
true
```

christiangnrd replied:
I've removed any sign of `atol` in the tests

giordano commented on test/fft.jl:
I may have missed such tests, but testing round-trip would also be good (`ifft(fft(x)) ≈ x`).

Also, they all work out-of-the-box already in this PR, but this is missing tests for `fft!`/`ifft!`/`bfft!`

christiangnrd replied:
Looking at the tests (borrowed from the CUDA fft tests), it seems like the heler functions (`complex_in_place` for example) test the forward pass, and then the inverse pass checks that it matches the original values, so I believe that covers round trip?

I've also added tests for `bfft` and `bfft!` with the "simple" `ifft` and `fft` tests.

giordano replied:
Ah, that's https://github.com/JuliaGPU/Metal.jl/blob/afbe65f610ea8128b4b1972a0f73baf64e38e72e/test/fft.jl#L129-L143?  Then yes, that should cover the round trip.

giordano commented on src/fft.jl:
```diff
@@ -0,0 +1,321 @@
+
+    function MtlFFTPlan{T, S, backward, inplace, N, R}(input_size::NTuple{N, Int}, output_size::NTuple{N, Int}, region::NTuple{R, Int}) where {T <: FFTNumber, S <: FFTNumber, backward, inplace, N, R}
+        # Validate region
+        if any(diff(collect(region)) .< 1)
```
Something like
```suggestion
        if any(i -> region[i] >= region[i+1], 1:R-1)
```
would avoid allocations

christiangnrd replied:
Done

giordano reviewed (commented):
I believe the AbstractFFT interface is implemented correctly, I mainly have some comments about the tests.

For the record, some benchmarks on M4
```julia-repl
julia> using BenchmarkTools, FFTW, Metal

julia> @btime plan * x setup=(x = randn(ComplexF32, 1024, 1024); plan = plan_fft!(x)) evals=1;
  14.898 ms (0 allocations: 0 bytes)

julia> @btime plan * x setup=(x = MtlArray(randn(ComplexF32, 1024, 1024)); plan = plan_fft!(x)) evals=1;
  224.041 μs (89 allocations: 3.08 KiB)

julia> @btime plan * x setup=(x = randn(ComplexF32, 4096, 4096); plan = plan_fft!(x)) evals=1;
  376.389 ms (0 allocations: 0 bytes)

julia> @btime plan * x setup=(x = MtlArray(randn(ComplexF32, 4096, 4096)); plan = plan_fft!(x)) evals=1;
  2.559 ms (89 allocations: 3.08 KiB)
```

christiangnrd committed:
Address feedback

giordano reviewed (approved):
I'm looking forward to playing with this feature 🚀

maleadt commented on lib/mpsgraphs/fft.jl:
```diff
@@ -0,0 +1,40 @@
+Create an MPSGraphFFTDescriptor with the specified parameters.
+"""
+function MPSGraphFFTDescriptor(; inverse::Bool = false, scalingMode::MPSGraphFFTScalingMode = MPSGraphFFTScalingModeNone)
+    obj = @objc [MPSGraphFFTDescriptor alloc]::id{MPSGraphFFTDescriptor}
```
Shouldn't this be `new`, or `alloc` + `init`?

christiangnrd replied:
[Fixed](https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraphfftdescriptor/descriptor?language=objc)

christiangnrd committed:
Use official API for the constructor
